go/arrow/flight/session/session.go

```diff
@@ -0,0 +1,233 @@
+type ServerSession interface {
+	// An identifier for the session that the server can use to reconstruct
+	// the session state on future requests. It is the responsibility of
+	// each implementation to define the token's semantics.
+	Token() string
+	// Get session option value by name, or nil if it does not exist
+	GetSessionOption(name string) *flight.SessionOptionValue
+	// Get a copy of the session options
+	GetSessionOptions() map[string]*flight.SessionOptionValue
+	// Set session option by name to given value
+	SetSessionOption(name string, value *flight.SessionOptionValue)
+	// Idempotently remove name from this session
+	EraseSessionOption(name string)
+	// Close the session
+	Close() error
+	// Report whether the session has been closed
+	Closed() bool
+}
+
+// Handles session lifecycle management
+type ServerSessionManager interface {
+	// Create a new, empty ServerSession
+	CreateSession(ctx context.Context) (ServerSession, error)
+	// Get the current ServerSession, if one exists
+	GetSession(ctx context.Context) (ServerSession, error)
+	// Cleanup any resources associated with the current ServerSession
+	CloseSession(session ServerSession) error
+}
+
+// Implementation of common session behavior. Intended to be extended
+// by specific session implementations.
+type serverSession struct {
+	closed bool
+
+	options map[string]*flight.SessionOptionValue
+	mu      sync.RWMutex
+}
+
+func (session *serverSession) GetSessionOption(name string) *flight.SessionOptionValue {
+	session.mu.RLock()
+	defer session.mu.RUnlock()
+	value, found := session.options[name]
+	if !found {
+		return nil
+	}
+
+	return value
+}
+
+func (session *serverSession) GetSessionOptions() map[string]*flight.SessionOptionValue {
+	options := make(map[string]*flight.SessionOptionValue, len(session.options))
+
+	session.mu.RLock()
+	defer session.mu.RUnlock()
+	for k, v := range session.options {
+		options[k] = proto.Clone(v).(*flight.SessionOptionValue)
+	}
+
+	return options
+}
+
+func (session *serverSession) SetSessionOption(name string, value *flight.SessionOptionValue) {
+	if value.GetOptionValue() == nil {
+		session.EraseSessionOption(name)
+		return
+	}
+
+	session.mu.Lock()
+	defer session.mu.Unlock()
+	session.options[name] = value
+}
+
+func (session *serverSession) EraseSessionOption(name string) {
+	session.mu.Lock()
+	defer session.mu.Unlock()
+	delete(session.options, name)
+}
+
+func (session *serverSession) Close() error {
+	session.options = nil
+	session.closed = true
+	return nil
+}
+
+func (session *serverSession) Closed() bool {
+	return session.closed
+}
+
+// Create new instance of CustomServerMiddleware implementing server session persistence.
+//
+// The provided manager can be used to customize session implementation/behavior.
+// If no manager is provided, a stateful in-memory, goroutine-safe implementation is used.
+func NewServerSessionMiddleware(manager ServerSessionManager) *serverSessionMiddleware {
+	// Default manager
```

Comment:
@lidavidm Was a default defined or specified for handling with FlightSQL sessions?

Reply:
No. It's up to the server implementation. The only requirement is that the client faithfully pass back cookies.

Reply:
The shipped middleware classes for C++ and Java both provide their own option value store, presently without the option to inject a separate k:v store implementation—although that could be accomplished via a subclass of the provided middleware factory.

Reply:
Okay given this I think this default here is fine. It's already in the docs explaining that if nil is provided they know what they will get back so i think we're good here. Anyone who wants to have a custom implementation could simply implement the `ServerSessionManager` interface themselves rather than using one of the ones we are providing.